## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#4fb77a53-ad5b-4ecf-ac7b-3113338a3d0a) |
| mint-gateway-azure.sh | :x: [more...](#e91be396-cf10-41dc-8e0d-be57e9bceee2) |
| mint-gateway-nas.sh | :x: [more...](#bb32cdaa-42a0-4795-85b1-df4980194660) |
<a name="4fb77a53-ad5b-4ecf-ac7b-3113338a3d0a"></a>
### 10434-fe11470/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev8.minio.io:30301
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c70371978665:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 3 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 44 seconds
(4/15) Running aws-sdk-ruby tests ... done in 5 seconds
(5/15) Running awscli tests ... FAILED in 12 minutes and 11 seconds
{"name": "awscli", "duration": 670206, "function": "aws --endpoint-url http://minio-dev8.minio.io:30301 s3 sync --no-progress /mint/data s3://awscli-mint-test-bucket-23066/\n", "status": "FAIL", "error": "upload: ../../../data/datafile-0-b to s3://awscli-mint-test-bucket-23066/datafile-0-bupload: ../../../data/datafile-1-kB to s3://awscli-mint-test-bucket-23066/datafile-1-kBupload: ../../../data/datafile-10-kB to s3://awscli-mint-test-bucket-23066/datafile-10-kBupload: ../../../data/datafile-100-kB to s3://awscli-mint-test-bucket-23066/datafile-100-kBupload: ../../../data/datafile-1-b to s3://awscli-mint-test-bucket-23066/datafile-1-bupload: ../../../data/datafile-1-MB to s3://awscli-mint-test-bucket-23066/datafile-1-MBupload: ../../../data/datafile-33-kB to s3://awscli-mint-test-bucket-23066/datafile-33-kBupload: ../../../data/datafile-1.03-MB to s3://awscli-mint-test-bucket-23066/datafile-1.03-MBupload: ../../../data/datafile-5243880-b to s3://awscli-mint-test-bucket-23066/datafile-5243880-bupload: ../../../data/datafile-6-MB to s3://awscli-mint-test-bucket-23066/datafile-6-MBupload: ../../../data/datafile-5-MB to s3://awscli-mint-test-bucket-23066/datafile-5-MBupload: ../../../data/datafile-10-MB to s3://awscli-mint-test-bucket-23066/datafile-10-MBupload: ../../../data/datafile-11-MB to s3://awscli-mint-test-bucket-23066/datafile-11-MBupload: ../../../data/datafile-129-MB to s3://awscli-mint-test-bucket-23066/datafile-129-MBupload failed: ../../../data/datafile-65-MB to s3://awscli-mint-test-bucket-23066/datafile-65-MB Connection was closed before we received a valid response from endpoint URL: "http://minio-dev8.minio.io:30301/awscli-mint-test-bucket-23066/datafile-65-MB?uploadId=fcc3edcc-707f-4360-9af7-dd0e65f401ad&partNumber=3"."}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 36 seconds
(7/15) Running minio-dotnet tests ... done in 40 seconds
(8/15) Running minio-go tests ... done in 1 minutes and 47 seconds
(9/15) Running minio-java tests ... done in 1 minutes and 16 seconds
(10/15) Running minio-js tests ... done in 51 seconds
(11/15) Running minio-py tests ... done in 3 minutes and 16 seconds
(12/15) Running s3cmd tests ... done in 21 seconds
(13/15) Running s3select tests ... done in 8 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="e91be396-cf10-41dc-8e0d-be57e9bceee2"></a>
### 10434-fe11470/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:31880
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 8adf3eda1fcf:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... FAILED in 5 minutes and 6 seconds
{
  "name": "aws-sdk-php",
  "function": "getBucketPolicy ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-5f5834e11934d"
  },
  "duration": "40633",
  "status": "FAIL",
  "error": "Error executing \"CreateBucket\" on \"http://minio-c3.minio.io:31880/aws-sdk-php-5f5834e11934d\"; AWS HTTP error: Client error: `PUT http://minio-c3.minio.io:31880/aws-sdk-php-5f5834e11934d` resulted in a `404 Not Found` response:\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>NoSuchBucket</Code><Message>The specified bucket does not exist</Mes (truncated...)\n NoSuchBucket (client): The specified bucket does not exist - <?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>NoSuchBucket</Code><Message>The specified bucket does not exist</Message><BucketName>aws-sdk-php-5f5834e11934d</BucketName><Resource>/aws-sdk-php-5f5834e11934d</Resource><RequestId>1632FAC2AF0E2620</RequestId><HostId>ac8d5ff5-8568-43b0-88e4-8c2f170db5ab</HostId></Error>"
}
(3/15) Running aws-sdk-ruby tests ... done in 19 seconds
(4/15) Running awscli tests ... FAILED in 45 seconds
{
  "name": "awscli",
  "duration": 3939,
  "function": "aws --endpoint-url http://minio-c3.minio.io:31880 s3api copy-object --bucket awscli-mint-test-bucket-21966 --key datafile-1-kB-copy --copy-source awscli-mint-test-bucket-21966/datafile-1-kB\n",
  "status": "FAIL",
  "error": "Hash mismatch expected a861ac626e0bee4421a50737c581233c, got 0x8D85463947C52B2-1"
}
(4/15) Running healthcheck tests ... done in 0 seconds
(5/15) Running mc tests ... done in 3 minutes and 19 seconds
(6/15) Running minio-dotnet tests ... done in 1 minutes and 40 seconds
(7/15) Running minio-go tests ... done in 5 minutes and 53 seconds
(8/15) Running minio-java tests ... done in 7 minutes and 44 seconds
(9/15) Running minio-js tests ... done in 2 minutes and 36 seconds
(10/15) Running minio-py tests ... FAILED in 13 minutes and 18 seconds
{
  "name": "minio-py:test_thread_safe",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-b0c9a74a-da7b-47c5-a731-63f46b88c820",
    "object_name": "4a6bdcc7-6c13-44e5-94dd-80a008178902"
  },
  "message": "InvalidSizeError: message: Data written 3145728 bytes is smaller than the specified size 11534336 bytes",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 147, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 1667, in test_thread_safe\n    raise exceptions[0]\n  File \"/mint/run/core/minio-py/tests.py\", line 1627, in get_object_and_check\n    _CLIENT.fget_object(bucket_name, object_name, local_file)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 822, in fget_object\n    raise InvalidSizeError(msg)\nminio.error.InvalidSizeError: InvalidSizeError: message: Data written 3145728 bytes is smaller than the specified size 11534336 bytes\n",
  "duration": 27427
}
(10/15) Running s3cmd tests ... done in 2 minutes and 9 seconds
(11/15) Running s3select tests ... done in 58 seconds
(12/15) Running security tests ... done in 0 seconds

Executed 12 out of 15 tests successfully.
```
<a name="bb32cdaa-42a0-4795-85b1-df4980194660"></a>
### 10434-fe11470/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:31438
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp e4077c843c41:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 1 minutes and 6 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... FAILED in 16 minutes and 3 seconds
{
  "name": "mc",
  "duration": "960985",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-9902 --quiet --no-color cp /mint/data/datafile-65-MB myminio/mc-test-bucket-10644/mc-test-object-7997  >>> \n`/mint/data/datafile-65-MB` -> `myminio/mc-test-bucket-10644/mc-test-object-7997`\nmc: <ERROR> Failed to copy `/mint/data/datafile-65-MB`. Bucket `mc-test-bucket-10644` does not exist.\nTotal: 0 B, Transferred: 19.94 MiB, Speed: 21.25 KiB/s"
}
(7/15) Running minio-dotnet tests ... done in 28 seconds
(8/15) Running minio-go tests ... done in 30 seconds
(9/15) Running minio-java tests ... done in 17 seconds
(10/15) Running minio-js tests ... done in 30 seconds
(11/15) Running minio-py tests ... done in 1 minutes and 7 seconds
(12/15) Running s3cmd tests ... done in 14 seconds
(13/15) Running s3select tests ... done in 2 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally
Error: No such image: minio/minio:10434-fe11470

--- a/docs/orchestration/docker-compose/README.md
+++ b/docs/orchestration/docker-compose/README.md
@@ -11,7 +11,7 @@ With Compose, you use a Compose file to configure MinIO services. Then, using a 
 
 ## 2. Run Distributed MinIO on Docker Compose
 
-To deploy Distributed MinIO on Docker Compose, please download [docker-compose.yaml](https://github.com/minio/minio/blob/master/docs/orchestration/docker-compose/docker-compose.yaml?raw=true) to your current working directory. Note that Docker Compose pulls the MinIO Docker image, so there is no need to explicitly download MinIO binary. Then run one of the below commands
+To deploy Distributed MinIO on Docker Compose, please download [docker-compose.yaml](https://github.com/minio/minio/blob/master/docs/orchestration/docker-compose/docker-compose.yaml?raw=true) and [nginx.conf](https://github.com/minio/minio/blob/master/docs/orchestration/docker-compose/nginx.conf?raw=true) to your current working directory. Note that Docker Compose pulls the MinIO Docker image, so there is no need to explicitly download MinIO binary. Then run one of the below commands
 
 ### GNU/Linux and macOS
 
@@ -27,7 +27,7 @@ docker-compose.exe pull
 docker-compose.exe up
 ```
 
-Each instance is now accessible on the host at ports 9001 through 9004, proceed to access the Web browser at http://127.0.0.1:9001/
+Distributed instances are now accessible on the host at ports 9000, proceed to access the Web browser at http://127.0.0.1:9000/. Here 4 MinIO server instances are reverse proxied through Nginx load balancing.
 
 ### Notes
 
@@ -36,11 +36,9 @@ Each instance is now accessible on the host at ports 9001 through 9004, proceed 
 * There are 4 minio distributed instances created by default. You can add more MinIO services (up to total 16) to your MinIO Compose deployment. To add a service
   * Replicate a service definition and change the name of the new service appropriately.
   * Update the command section in each service.
-  * Update the port number to exposed for the new service. Also, make sure the port assigned for the new service is not already being used on the host.
+  * Add a new MinIO server instance to the upstream directive in the Nginx configuration file.
 
   Read more about distributed MinIO [here](https://docs.min.io/docs/distributed-minio-quickstart-guide).
-
-* MinIO services in the Docker compose file expose ports 9001 to 9004. This allows multiple services to run on a host.
 
 ### Explore Further
 - [Overview of Docker Compose](https://docs.docker.com/compose/overview/)

--- a/docs/orchestration/docker-compose/docker-compose.yaml
+++ b/docs/orchestration/docker-compose/docker-compose.yaml
@@ -1,16 +1,16 @@
 version: '3.7'
 
-# starts 4 docker containers running minio server instances. Each
-# minio server's web interface will be accessible on the host at port
-# 9001 through 9004.
+# starts 4 docker containers running minio server instances.
+# using nginx reverse proxy, load balancing, you can access
+# it through port 9000.
 services:
   minio1:
     image: minio/minio:RELEASE.2020-09-08T23-05-18Z
     volumes:
       - data1-1:/data1
       - data1-2:/data2
-    ports:
-      - "9001:9000"
+    expose:
+      - "9000"
     environment:
       MINIO_ACCESS_KEY: minio
       MINIO_SECRET_KEY: minio123
@@ -26,8 +26,8 @@ services:
     volumes:
       - data2-1:/data1
       - data2-2:/data2
-    ports:
-      - "9002:9000"
+    expose:
+      - "9000"
     environment:
       MINIO_ACCESS_KEY: minio
       MINIO_SECRET_KEY: minio123
@@ -43,8 +43,8 @@ services:
     volumes:
       - data3-1:/data1
       - data3-2:/data2
-    ports:
-      - "9003:9000"
+    expose:
+      - "9000"
     environment:
       MINIO_ACCESS_KEY: minio
       MINIO_SECRET_KEY: minio123
@@ -60,8 +60,8 @@ services:
     volumes:
       - data4-1:/data1
       - data4-2:/data2
-    ports:
-      - "9004:9000"
+    expose:
+      - "9000"
     environment:
       MINIO_ACCESS_KEY: minio
       MINIO_SECRET_KEY: minio123
@@ -71,6 +71,18 @@ services:
       interval: 30s
       timeout: 20s
       retries: 3
+
+  nginx:
+    image: nginx:1.19.2-alpine
+    volumes:
+      - ./nginx.conf:/etc/nginx/nginx.conf:ro
+    ports:
+      - "9000:9000"
+    depends_on:
+      - minio1
+      - minio2
+      - minio3
+      - minio4
 
 ## By default this config uses default local driver,
 ## For custom volumes replace with volume driver configuration.

--- a/docs/orchestration/docker-compose/nginx.conf
+++ b/docs/orchestration/docker-compose/nginx.conf
@@ -1,0 +1,68 @@
+
+user  nginx;
+worker_processes  auto;
+
+error_log  /var/log/nginx/error.log warn;
+pid        /var/run/nginx.pid;
+
+
+events {
+    worker_connections  1024;
+}
+
+
+http {
+    include       /etc/nginx/mime.types;
+    default_type  application/octet-stream;
+
+    log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
+                      '$status $body_bytes_sent "$http_referer" '
+                      '"$http_user_agent" "$http_x_forwarded_for"';
+
+    access_log  /var/log/nginx/access.log  main;
+
+    sendfile        on;
+    #tcp_nopush     on;
+
+    keepalive_timeout  65;
+
+    #gzip  on;
+
+    # include /etc/nginx/conf.d/*.conf;
+
+    upstream minio {
+        server minio1:9000;
+        server minio2:9000;
+        server minio3:9000;
+        server minio4:9000;
+    }
+
+    server {
+        listen       9000;
+        listen  [::]:9000;
+        server_name  localhost;
+
+         # To allow special characters in headers
+         ignore_invalid_headers off;
+         # Allow any size file to be uploaded.
+         # Set to a value such as 1000m; to restrict file size to a specific value
+         client_max_body_size 0;
+         # To disable buffering
+         proxy_buffering off;
+
+        location / {
+            proxy_set_header Host $http_host;
+            proxy_set_header X-Real-IP $remote_addr;
+            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+            proxy_set_header X-Forwarded-Proto $scheme;
+
+            proxy_connect_timeout 300;
+            # Default is HTTP/1, keepalive is only enabled in HTTP/1.1
+            proxy_http_version 1.1;
+            proxy_set_header Connection "";
+            chunked_transfer_encoding off;
+
+            proxy_pass http://minio;
+        }
+    }
+}

